### PR TITLE
[Merged by Bors] - add aggregated layer hash in layer hash response

### LINKED ISSUE #2524: add aggregated hashes in layer response
## Description
part of #2507

currently nodes exchange simple layer hashes. if two nodes simply exchange layer hashes for the past X layers, there may be multiple layers with different hashes and one is still none the wiser about where the point of divergence is. the only way is to exchange aggregated layer hashes where 

aggregatedHash(N) = hash(aggregatedHash(N-1) + hash(N)) 

to ensure there is a point of divergence between two peers.

### PR DESCRIPTION
**Pull request by countvonzero**
## Motivation
Closes #2524

## Changes
- only persist layer hash after validity is determined
- add aggregated layer hash and processed layer in layer hash response

## Test Plan
unit tests

## DevOps Notes

- [x] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [x] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [x] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

### CODE REVIEW

**Comment by countvonzero** [comment]
bors try

File: mesh/mesh_test.go [hunk 1]
```diff
@@ -163,6 +163,58 @@ func addLayer(r *require.Assertions, id types.LayerID, layerSize int, msh *Mesh)
...
+	err = msh.SetZeroBlockLayer(lyrID)
+	assert.NoError(t, err)
+	assert.Equal(t, EmptyLayerHash, msh.GetLayerHash(lyrID))
+
+	// it's ok to add to an empty layer
+	lyr = addLayer(r, lyrID, 10, msh)
+	msh.ValidateLayer(lyr)
+	lyr, err = msh.GetLayer(lyrID)
+	assert.NoError(t, err)
+	assert.Equal(t, 10, len(lyr.Blocks()))
```
Comment (dshulyak) [inline mesh/mesh_test.go:235]: this will fail with nil pointer dereference if err is not nil. i think it is desirable to use require.Smth if code below the statement will crash if the assertion didn't pass

File: layerfetcher/layers.go [hunk 2]
```diff
@@ -342,13 +356,18 @@ func notifyLayerHashResult(layerID types.LayerID, channels []chan LayerHashResul
...
 	for peer, res := range peerResult {
 		if res.err != nil {
 			numErrors++
 		} else {
-			hash := types.BytesToHash(res.data)
-			hashes[hash] = append(hashes[hash], peer)
+			var lyrHash layerHash
+			convertErr := types.BytesToInterface(res.data, &lyrHash)
+			if convertErr != nil {
+				logger.Error("received error converting bytes to layerHash", convertErr)
```
Comment (dshulyak) [inline layerfetcher/layers.go]: we have a popup notification for every error in the code. could you maybe change it to warn or debug? i personally think that any data that was received from the network and can't be unmarshalled is a debug information

File: layerfetcher/layers.go [hunk 3]
```diff
@@ -192,11 +193,24 @@ func (l *Logic) AddDBs(blockDB, AtxDB, TxDB, poetDB, IvDB, tbDB database.Getter)
...
-		log.String("layerHash", h.ShortString()))
-	return h.Bytes()
+	lyrHash := &layerHash{
+		ProcessedLayer: l.layerDB.ProcessedLayer(),
+		SimpleHash:     l.layerDB.GetLayerHash(lyr),
+		AggHash:        l.layerDB.GetAggregatedLayerHash(lyr),
+	}
+	out, err := types.InterfaceToBytes(lyrHash)
+	if err != nil {
+		l.log.WithContext(ctx).With().Error("failed to serialize layer hash",
```
Comment (dshulyak) [inline layerfetcher/layers.go]: Panic is cleaner here. InterfaceToBytes returns error only if the input is incorrect (and in this case it is formed by the developer)

File: mesh/mesh_test.go [hunk 4]
```diff
@@ -163,6 +163,58 @@ func addLayer(r *require.Assertions, id types.LayerID, layerSize int, msh *Mesh)
...
+	// after a layer is validated, only contextually valid blocks count towards its hash
+	expHash := types.CalcBlocksHash32(types.SortBlockIDs(types.BlockIDs(lyr.Blocks()[1:])), nil)
+	assert.Equal(t, expHash, msh.GetLayerHash(lyrID))
+	assert.NotEqual(t, expHash, lyr.Hash())
+}
+
+func TestMesh_SetZeroBlockLayer(t *testing.T) {
+	r := require.New(t)
+	msh := getMesh(t, "zero block layer")
+	defer msh.Close()
```
Comment (dshulyak) [inline mesh/mesh_test.go]: t.Cleanup(msh.Close)

File: mesh/mesh_test.go [hunk 5]
```diff
@@ -163,6 +163,58 @@ func addLayer(r *require.Assertions, id types.LayerID, layerSize int, msh *Mesh)
 	return l
 }
 
+func TestMesh_GetLayerHash(t *testing.T) {
+	r := require.New(t)
+	msh := getMesh(t, "get layer hash")
+	defer msh.Close()
```
Comment (dshulyak) [inline mesh/mesh_test.go]: t.Cleanup

File: layerfetcher/wire_types.go [hunk 6]
```diff
@@ -2,7 +2,17 @@ package layerfetcher
...
 
-// layerBlocks is the response for layer hash
+// layerHash is the response for a given layer
+type layerHash struct {
+	// ProcessedLayer is the latest processed layer from peer
+	ProcessedLayer types.LayerID
+	// SimpleHash is the hash of contextually valid blocks (sorted by block ID) in the given layer
+	SimpleHash types.Hash32
+	// AggHash is the aggregated hash of all layers up to the given layer
+	AggHash types.Hash32
```
Comment (dshulyak) [inline layerfetcher/wire_types.go]: could it be a full word please?

File: layerfetcher/wire_types.go [hunk 7]
```diff
@@ -2,7 +2,17 @@ package layerfetcher
 
 import "github.com/spacemeshos/go-spacemesh/common/types"
 
-// layerBlocks is the response for layer hash
+// layerHash is the response for a given layer
+type layerHash struct {
+	// ProcessedLayer is the latest processed layer from peer
+	ProcessedLayer types.LayerID
+	// SimpleHash is the hash of contextually valid blocks (sorted by block ID) in the given layer
+	SimpleHash types.Hash32
```
Comment (dshulyak) [inline layerfetcher/wire_types.go]: i think just Hash is easier to understand compared with SimpleHash

File: layerfetcher/wire_types.go [hunk 8]
```diff
@@ -2,7 +2,17 @@ package layerfetcher
 
 import "github.com/spacemeshos/go-spacemesh/common/types"
 
-// layerBlocks is the response for layer hash
+// layerHash is the response for a given layer
```
Comment (dshulyak) [inline layerfetcher/wire_types.go]: it seems like a copy from another struct. better to remove it altogether

**Review by dshulyak — CHANGES_REQUESTED** [review]

**Comment by countvonzero** [comment]
bors try

**Comment by countvonzero** [comment]
bors try

File: mesh/mesh.go [hunk 9]
```diff
@@ -236,6 +236,22 @@ func (msh *Mesh) GetLayer(i types.LayerID) (*types.Layer, error) {
...
 
+// GetLayerHash returns layer hash for received blocks
+func (msh *Mesh) GetLayerHash(layerID types.LayerID) types.Hash32 {
+	h, err := msh.recoverLayerHash(layerID)
+	if err == nil {
+		return h
+	}
+	if err == database.ErrNotFound {
+		// layer hash not persisted. i.e. contextual validity not yet determined
+		lyr, err := msh.GetLayer(layerID)
```
Comment (antonlerner) [inline mesh/mesh.go:247]: this means that layer hash might be changed when layer validity is determined, is this wanted behaviour?

File: mesh/mesh.go [hunk 10]
```diff
@@ -575,6 +568,15 @@ func (msh *Mesh) persistAggregatedLayerHash(layerID types.LayerID, hash types.Ha
 	}
 }
 
+// GetAggregatedLayerHash returns the aggregated layer hash up to the specified layer
+func (msh *Mesh) GetAggregatedLayerHash(layerID types.LayerID) types.Hash32 {
+	h, err := msh.getAggregatedLayerHash(layerID)
+	if err != nil {
+		return EmptyLayerHash
```
Comment (antonlerner) [inline mesh/mesh.go:575]: shouldn't this return an error? event if the requested layer has no blocks the aggregated hash shouldn't be zero

File: mesh/mesh.go (hunk 9 above)
Comment (countvonzero) [reply mesh/mesh.go:247]: actually the behavior in this area is mostly unchanged by this PR.

- before this change:
when we received blocks via gossip, the code persists the layer hash as the blocks are received. see 
https://github.com/spacemeshos/go-spacemesh/blob/develop/mesh/meshdb.go#L436
before the layer is validated, it simply calculates the hash of all blocks since validity isn't determined yet. but after a layer is validated, a problem can rise when we receive a late block. because the code immediately persists the layer hash when we get a new block, before it determines whether the block is valid or not.

in short, the persisted hash **may** contains invalid blocks.

- after this change
layer hash is only persisted **after** validation happens. before the validation runs for a given layer, no hash is persisted. that's why i added the code here to simply return hash of all blocks. no validity has been determined. after the layer is validated, then we only let tortoise update the layer hash as it deem a block valid.

in short, the persisted hash **only** contains valid blocks.

in both cases, layer hash can change from `all blocks` to only `valid blocks` after validation.

File: mesh/mesh.go (hunk 10 above)
Comment (countvonzero) [reply mesh/mesh.go:575]: exactly. EmptyLayerHash is actually the way to indicate error for aggregated layer. it will **never** be EmptyLayerHash, even for layer 1.

to be honest i was trying to pass around an error between peers but xdr encoding doesn't work with nil 
`xdr:encodeInterface: can't encode nil interface`

File: layerfetcher/layers.go (hunk 3 above)
Comment (countvonzero) [reply layerfetcher/layers.go]: done. thanks

File: layerfetcher/layers.go (hunk 2 above)
Comment (countvonzero) [reply layerfetcher/layers.go]: what is your opinion in the case of DDoS attack? i guess in this case we have metrics (numErrors). but in general i think warning may be appropriate. changed to debug for now.

File: layerfetcher/wire_types.go (hunk 8 above)
Comment (countvonzero) [reply layerfetcher/wire_types.go]: done

File: layerfetcher/wire_types.go (hunk 7 above)
Comment (countvonzero) [reply layerfetcher/wire_types.go]: done

File: layerfetcher/wire_types.go (hunk 6 above)
Comment (countvonzero) [reply layerfetcher/wire_types.go]: done

File: mesh/mesh_test.go (hunk 5 above)
Comment (countvonzero) [reply mesh/mesh_test.go]: done for the whole test

File: mesh/mesh_test.go (hunk 4 above)
Comment (countvonzero) [reply mesh/mesh_test.go]: done

File: mesh/mesh_test.go (hunk 1 above)
Comment (countvonzero) [reply mesh/mesh_test.go:235]: thanks. done. and other places (i can find anyway. eyes got blurry after a while)

**Review by dshulyak — APPROVED** [review]

**Review by antonlerner — APPROVED** [review]

**Comment by countvonzero** [comment]
bors merge

**Comment by countvonzero** [comment]
bors merge
